src/core/timer.c

```diff
@@ -394,6 +394,15 @@ static void timer_enter_waiting(Timer *t, bool time_change) {
                 if (v->base == TIMER_CALENDAR) {
                         usec_t b, rebased;
 
+                        /* Update last_trigger to 'now' in case
+                         * the system time changes, so that
+                         * next_elapse is not stuck with a
+                         * future date. */
+                        if (time_change){
+                                t->last_trigger.realtime = ts.realtime;
```

Comment:
This sounds hackish for me. This intends to set `b = ts.realtime`, no? How about to update the below condition like the following?
```c
if (time_change)
        b = ts.realtime;
else if (t->last_trigger.realtime > 0)
        b = t->last_trigger.realtime;
else if (state_translation_table[t->state] == UNIT_ACTIVE)
        b = UNIT(t)->inactive_exit_timestamp.realtime;
else
        b = ts.realtime;
```

Reply:
I gave a try, and the tests fails with the patch you suggested. I will give it a thought and try to improve on it. 
Here are the values for the failing test:
```bash
[    2.739706] testsuite-53.sh[317]: + next_elapsed_t1='Sat 2022-08-06 22:45:00 CEST'
[    2.741909] testsuite-53.sh[330]: ++ date -d 'Sat 2022-08-06 22:45:00 CEST' +%s
[    2.743462] testsuite-53.sh[317]: + next_elapsed_t1=1659818700
[    2.746914] testsuite-53.sh[331]: ++ date +%s
[    2.749176] testsuite-53.sh[317]: + now=1659559136
[    2.750483] testsuite-53.sh[317]: + time_delta_t1=259564
[    2.770859] testsuite-53.sh[317]: + (( 0 < time_delta_t1 && time_delta_t1 < 1200 ))
[    2.772332] testsuite-53.sh[317]: + echo 'Timer elapse outside of the expected 20 minute window.'
[    2.773895] testsuite-53.sh[317]: + echo '  next_elapsed_t1=1659818700'
[    2.774993] testsuite-53.sh[317]: + echo '  now=1659559136'
[    2.776125] testsuite-53.sh[317]: + echo '  time_delta_t1=259564'

```